core/coreapi/interface/options/unixfs.go

```diff
@@ -0,0 +1,187 @@
+)
+
+type Layout int
+
+const (
+	BalancedLayout Layout = iota
+	TrickleLayout
+)
+
+type UnixfsAddSettings struct {
+	CidVersion int
+	MhType     uint64
+
+	InlineLimit  int
+	RawLeaves    bool
+	RawLeavesSet bool
+
+	Chunker string
+	Layout  Layout
+
+	Pin      bool
+	OnlyHash bool
+	Local    bool
+}
+
+type UnixfsAddOption func(*UnixfsAddSettings) error
+
+func UnixfsAddOptions(opts ...UnixfsAddOption) (*UnixfsAddSettings, cid.Prefix, error) {
+	options := &UnixfsAddSettings{
+		CidVersion: -1,
+		MhType:     mh.SHA2_256,
+
+		InlineLimit:  0,
+		RawLeaves:    false,
+		RawLeavesSet: false,
+
+		Chunker: "size-262144",
+		Layout:  BalancedLayout,
+
+		Pin:      false,
+		OnlyHash: false,
+		Local:    false,
+	}
+
+	for _, opt := range opts {
+		err := opt(options)
+		if err != nil {
+			return nil, cid.Prefix{}, err
+		}
+	}
+
+	// (hash != "sha2-256") -> CIDv1
+	if options.MhType != mh.SHA2_256 {
+		switch options.CidVersion {
+		case 0:
+			return nil, cid.Prefix{}, errors.New("CIDv0 only supports sha2-256")
+		case 1, -1:
+			options.CidVersion = 1
+		default:
+			return nil, cid.Prefix{}, fmt.Errorf("unknown CID version: %d", options.CidVersion)
+		}
+	} else {
+		if options.CidVersion < 0 {
+			// Default to CIDv0
+			options.CidVersion = 0
+		}
+	}
+
+	// cidV1 -> raw blocks (by default)
+	if options.CidVersion > 0 && !options.RawLeavesSet {
+		options.RawLeaves = true
+	}
+
+	prefix, err := dag.PrefixForCidVersion(options.CidVersion)
+	if err != nil {
+		return nil, cid.Prefix{}, err
+	}
+
+	prefix.MhType = options.MhType
+	prefix.MhLength = -1
+
+	return options, prefix, nil
+}
+
+type unixfsOpts struct{}
+
+var Unixfs unixfsOpts
+
+// CidVersion specifies which CID version to use. Defaults to 0 unless an option
+// that depends on CIDv1 is passed.
+func (unixfsOpts) CidVersion(version int) UnixfsAddOption {
+	return func(settings *UnixfsAddSettings) error {
+		settings.CidVersion = version
+		return nil
+	}
+}
+
+// Hash function to use. Implies CIDv1 if not set to sha2-256 (default).
+//
+// Table of functions is declared in https://github.com/multiformats/go-multihash/blob/master/multihash.go
+func (unixfsOpts) Hash(mhtype uint64) UnixfsAddOption {
+	return func(settings *UnixfsAddSettings) error {
+		settings.MhType = mhtype
+		return nil
+	}
+}
+
+// RawLeaves specifies whether to use raw blocks for leaves (data nodes with no
+// links) instead of wrapping them with unixfs structures.
+func (unixfsOpts) RawLeaves(enable bool) UnixfsAddOption {
+	return func(settings *UnixfsAddSettings) error {
+		settings.RawLeaves = enable
+		settings.RawLeavesSet = true
+		return nil
+	}
+}
+
+// InlineLimit sets the amount of bytes below which blocks will be encoded
+// directly into CID instead of being stored and addressed by it's hash
+//
+// Note that while there is no hard limit on the number of bytes here, it should
+// be kept at something reasonably low like 32b (default for 'ipfs add')
```

Comment:
The wording here makes it sounds like the value should be 32 bytes or smaller.  That is misleading.  Values less than 32 bytes should not be encouraged.  The default of 32 was chosen so that when the Inline option is enabled it won't increase the size of the hashes, but a limit of 32 bytes is not very usefel.  A more reasonable upper limit would be 64 bytes, but if you are not concerned with displaying it than values like 128 or even 256 are also okay and based on my informal testing do not cause problems anywhere in IPFS.